### PR TITLE
feat: Add layout configuration to exclude panes from tab sync

### PR DESCRIPTION
**Pull request by on3iro**
Closes: #2285 

This PR adds the ability to configure panes to be excluded from receiving input when tab sync is activated.
I'm not sure if there is a better way than to add another parameter to the `write_to_pane_id`-function - so just let me know if I should change something.

I also did not find any tests in regard to tab syncing, so I also didn't add any.

If this PR gets merged I should probably also make an addition to the zellij website providing an example in the layouts section. I am just a bit unsure right now, how these two repos are being kept in sync. 
Should I already make the PR or should I wait until this PR might've been merged?

### CODE REVIEW

**Review by imsnif — CHANGES_REQUESTED** [review]
Hey @on3iro - this is great work. Good job threading the attribute, I imagine it was not trivial to figure out!

I left some comments in the code, and also I think there's a bit of a functionality issue. Using this layout:
```kdl
layout {
    pane
    pane split_direction="vertical" {
        pane
        pane exclude_from_sync=true
    }
}

```
If I start tab-sync, I can't write into the excluded pane even if I'm focused on it. Would you like to take a look?

**Comment by on3iro** [comment]
> Hey @on3iro - this is great work. Good job threading the attribute, I imagine it was not trivial to figure out!
> 
> I left some comments in the code, and also I think there's a bit of a functionality issue. Using this layout:
> 
> ```
> layout {
>     pane
>     pane split_direction="vertical" {
>         pane
>         pane exclude_from_sync=true
>     }
> }
> ```
> 
> If I start tab-sync, I can't write into the excluded pane even if I'm focused on it. Would you like to take a look?

@imsnif thanks a lot for your kind words and feedback. I'll take a closer look at all the comments asap.
Regarding the functionality issue: I was aware of that behavior and it was a actually intentional on my side.
To me this behavior feels consistent, meaning that if I am inside sync mode, the pane should not receive input, no matter where I type. However I totally understand if you don't desire such behavior.
Should I try to make it so that the pane will always receive input when it is in focus? :)

**Comment by imsnif** [comment]
> Should I try to make it so that the pane will always receive input when it is in focus? :)

Actually, this makes sense. Let's leave it like this.

File: zellij-utils/src/input/unit/layout_test.rs [hunk 1]
```diff
@@ -1008,6 +1008,17 @@ fn combined_tab_and_pane_template_both_with_children() {
...
 
+#[test]
+fn layout_with_pane_excluded_from_sync() {
+    let kdl_layout = r#"
+        layout {
+            pane exclude_from_sync=true
+        }
+    "#;
+    let layout = Layout::from_kdl(kdl_layout, "layout_file_name".into(), None, None).unwrap();
+    assert_snapshot!(format!("{:?}", layout));
```
[the remark this answers is not in the document]
Comment (on3iro) [reply zellij-utils/src/input/unit/layout_test.rs:1019]: sure, will do :)

File: zellij-server/src/tab/mod.rs [hunk 2]
```diff
@@ -1543,12 +1551,18 @@ impl Tab {
...
+                        if !is_synced || !active_terminal.exclude_from_sync() {
+                            self.senders
+                                .send_to_pty_writer(PtyWriteInstruction::Write(
+                                    adjusted_input,
+                                    active_terminal_id,
+                                ))
+                                .with_context(err_context)?;
+                        }
                     },
                     Some(AdjustedInput::ReRunCommandInThisPane(command)) => {
```
[the remark this answers is not in the document]
Comment (on3iro) [reply zellij-server/src/tab/mod.rs:1567]: Oh yeah. Good point. I totally missed that :+1:

File: zellij-server/src/tab/mod.rs [hunk 3]
```diff
@@ -1543,12 +1551,18 @@ impl Tab {
...
-                            .send_to_pty_writer(PtyWriteInstruction::Write(
-                                adjusted_input,
-                                active_terminal_id,
-                            ))
-                            .with_context(err_context)?;
+                        // We always write for non-synced terminals.
+                        // However if the terminal is part of a tab-sync, we need to
+                        // check if the terminal should receive input or not (depending on its
+                        // 'exclude_from_sync' configuration).
+                        if !is_synced || !active_terminal.exclude_from_sync() {
```
[the remark this answers is not in the document]
Comment (on3iro) [reply zellij-server/src/tab/mod.rs]: Great idea!

File: zellij-server/src/tab/mod.rs [hunk 4]
```diff
@@ -1520,14 +1523,19 @@ impl Tab {
...
         }
         Ok(())
     }
 
-    pub fn write_to_pane_id(&mut self, input_bytes: Vec<u8>, pane_id: PaneId) -> Result<bool> {
+    pub fn write_to_pane_id(
+        &mut self,
+        input_bytes: Vec<u8>,
+        pane_id: PaneId,
+        is_synced: bool,
```
[the remark this answers is not in the document]
Comment (on3iro) [reply zellij-server/src/tab/mod.rs]: Yeah that makes sense. I somehow totally missed that I can just call the method and don't really need to drill the property down to the write method.

**Comment by on3iro** [comment]
@imsnif I made the changes you requested. I kept them inside a separate commit for now, so they are easier to distinguish from the previous code. If you are satisfied with them you can just squash 'em :) 

Just let me know if there is anything else I should change :+1:

**Comment by imsnif** [comment]
Hey @on3iro - the code changes look good, thanks! If you'd like to merge from main again, we can give this one last spin to make sure all is working and then merge. I'll try to get to it quickly after you merge in order not to create more drift!

**Comment by on3iro** [comment]
Hey @imsnif - sorry for the late response. I am in the midst of moving into a new flat and super busy right now.
I will get back to this as soon as things have settled a bit. (probably around Thursday next week)

**Comment by on3iro** [comment]
@imsnif I squashed my two branches and rebased onto `main` - everything should be fine and ready to merge now :) 🎉

**Comment by imsnif** [comment]
Very nice work on this! I hope the move went well. Thank you for your patience.

**Pull request merged**